docs/reference/rego-templates.md

```diff
@@ -0,0 +1,214 @@
+          "isSuccess": true,
+          "message": "",
+          "extensions": {}
+        }
+      ],
+      "nestedReports": [
+        {
+          "artifactType": "application/vnd.cncf.notary.signature",
+          "subject": "test.azurecr.io/test/hello-world@sha256:123",
+          "referenceDigest": "test.azurecr.io/test/hello-world@sha256:33333",
+          "verifierReports": [
+            {
+              "verifierName": "notation",
+              "verifierType": "verifier-notation",
+              "isSuccess": true,
+              "message": "",
+              "extensions": {
+                "Issuer": "testIssuer",
+                "SN": "testSN"
+              }
+            }
+          ],
+          "nestedReports": []
+        }
+      ]
+    }  
+  ]
+}
+```
+
+### Example 1
+Require all reference artifacts associated with subject image to be verify successfully.
+Check equivalent [Config Policy](./providers.md#config-policy-examples)
+
+```rego
+package ratify.policy
+default valid := false
+valid {
+  not failed_verify(input)
+}
+failed_verify(reports) {
+  [path, value] := walk(reports)
+  value == false
+  path[count(path) - 1] == "isSuccess"
+}
+failed_verify(reports) {
+  [path, value] := walk(reports)
+  path[count(path) - 1] == "verifierReports"
+  count(value) == 0
+}
+```
+
+### Example 2
+Require at least one reference artifact of the same type to verify succesfully. (relaxes the default policy to 'any').
+Check equivalent [Config Policy](./providers.md#config-policy-examples)
+
+```rego
+package ratify.policy
+default valid := false
+valid if {
+	not failed_verify(input)
+}
+failed_verify(reports) if {
+	newReports := {"nestedReports": reports.verifierReports}
+	has_subject_failed_verify(newReports)
+}
+has_subject_failed_verify(nestedReports) if {
+	[path, value] := walk(nestedReports)
+	not artifact_type_pass_verify(value)
+	path[count(path) - 1] == "nestedReports"
+}
+# at least one artifact of the same type passed verification
+artifact_type_pass_verify(nestedReports) if {
+	count_artifact_type(nestedReports) == count_successful_artifact_type(nestedReports)
+}
+count_artifact_type(nestedReports) := number if {
+	artifact_types := {x |
+		some i
+		x := nestedReports[i].artifactType
+	}
+	number := count(artifact_types)
+}
+count_successful_artifact_type(nestedReports) := number if {
+	artifact_types := {x |
+		some i
+		x := nestedReports[i].artifactType
+		artifact_pass_verify(nestedReports[i].verifierReports)
+	}
+	number := count(artifact_types)
+}
+# an artifact has at least one successful report
+artifact_pass_verify(verifierReports) if {
+	verifierReports[_].isSuccess == true
+}
+# should be at least one verifier report that is successful
+failed_verify(reports) if {
+	[path, value] := walk(reports)
+	path[count(path) - 1] == "verifierReports"
+	count(value) == 0
+}
+```
+
+### Example 3
+I trust multiple identities, and need to find at least one valid signature signed by any identity.
```

Comment:
Q:  is example3 and 4 the same ? "I trust multiple identities, and need to find at least one valid signature signed by any identity."

Reply:
No, they're different. example 3 is `by any identity` while example 4 is `by each identity`. In example 3, only one signature is required, signed by any of the available identities. Conversely, in example 4, signatures must be signed by all trusted identities.
Do you have any suggestion to rephrase it?